internals/secrethub/tpl/v2.go

```diff
@@ -24,95 +26,439 @@ func NewV2Parser() Parser {
+//   variable tag is enclosed between ${ and }: `${ variable }`.
+// - Extra spaces can be added just after the opening delimiter and just before the
+//   closing delimiter of a tag: {{ path/to/secret }} has the same output as
+//   {{path/to/secret}} has.
+// - Secret tags can also contain variable tags: `{{ path/with/${var}/to/secret }}`
+// - Variable tags cannot contain secret tags.
+// - Secret tags cannot contain secret tags (they cannot be nested).
+// - Variable tags cannot contain variable tags (they cannot be nested).
+func (p parserV2) Parse(raw string, line, column int) (Template, error) {
+	parser := newV2Parser(bytes.NewBufferString(raw), line, column)
+
+	nodes, err := parser.parse()
 	if err != nil {
 		return nil, err
 	}
 
-	keys := t.Keys()
+	return templateV2{
+		nodes: nodes,
+	}, nil
+}
 
-	secrets := make(map[string]tpl.Template, len(keys))
+func newV2Parser(buf *bytes.Buffer, line, column int) v2Parser {
+	return v2Parser{
+		buf:    buf,
+		lineNo: line,
+		// The column number indicates the index (starting at 1) of the current rune.
+		// We subtract 2 of the given value. One because we have not read the current rune yet and
+		// one more because we are reading the next rune in advance (which we don't want to count).
+		columnNo: column - 2,
+	}
+}
+
+type v2Parser struct {
+	buf      *bytes.Buffer
+	lineNo   int
+	columnNo int
 
-	templateVarParser := tpl.NewParser("${", "}")
-	for _, k := range keys {
-		parsed, err := templateVarParser.Parse(k)
+	current rune
+	next    rune
+}
+
+// readRune reads the next rune from the raw template.
+func (p *v2Parser) readRune() error {
+	p.current = p.next
+	if p.current == '\n' {
+		p.lineNo++
+		p.columnNo = 0
+	} else {
+		p.columnNo++
+	}
+
+	var err error
+	p.next, _, err = p.buf.ReadRune()
+	return err
+}
+
+func (p *v2Parser) parse() ([]node, error) {
+	res := []node{}
+
+	err := p.readRune()
+	if err == io.EOF {
+		return res, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	for {
+		err := p.readRune()
+		if err == io.EOF {
+			return append(res, character(p.current)), nil
+		}
 		if err != nil {
 			return nil, err
 		}
-		secrets[k] = parsed
-	}
 
-	return templateV2{
-		template: t,
-		secrets:  secrets,
-	}, nil
+		switch p.current {
+		case token.Dollar:
+			switch p.next {
+			case token.LBracket:
+				variable, err := p.parseVar()
+				if err != nil {
+					return nil, err
+				}
+
+				res = append(res, variable)
+
+				err = p.readRune()
+				if err == io.EOF {
+					return res, nil
+				}
+				if err != nil {
+					return nil, err
+				}
+
+				continue
+			default:
+				// We don't allow dollars before letters and underscores now,
+				// as we might want to use these for $var support (without brackets) later.
+				if unicode.IsLetter(p.next) || p.next == '_' {
+					return nil, ErrUnexpectedDollar(p.lineNo, p.columnNo)
+				}
+
+				res = append(res, character(p.current))
+				continue
```

Comment:
I see a lot of these that are technically not necessary. I understand why you've added them: to force a break from this switch. So this made me think: can't this switch live in a function? That way, you can force the switch to end at the end of every case by returning. The compiler can then check that you haven't missed any returns (because you dont place a return at the end of the function, so every individual case must have a a return).

I can't easily see whether it is possible to extract this to a function without having to make the code ugly. Do you have any idea?

Reply:
Tried it and came up with this. What do you think?
```go
func (p *v2Parser) parse() ([]node, error) {
	res := []node{}

	err := p.readRune()
	if err == io.EOF {
		return res, nil
	}
	if err != nil {
		return nil, err
	}

	for {
		err := p.readRune()
		if err == io.EOF {
			return append(res, character(p.current)), nil
		}
		if err != nil {
			return nil, err
		}
		r, err := p.parseRune()
		if err == io.EOF {
			return append(res, r), nil
		}
		if err != nil {
			return nil, err
		}
		res = append(res, r)
	}
}

func (p *v2Parser) parseRune() (node, error) {
	switch p.current {
	case token.Dollar:
		switch p.next {
		case token.LBracket:
			variable, err := p.parseVar()
			if err != nil {
				return nil, err
			}
			return variable, p.readRune()
		default:
			// We don't allow dollars before letters and underscores now,
			// as we might want to use these for $var support (without brackets) later.
			if unicode.IsLetter(p.next) || p.next == '_' {
				return nil, ErrUnexpectedDollar(p.lineNo, p.columnNo)
			}
			return character(p.current), nil
		}
	case token.LBracket:
		switch p.next {
		case token.LBracket:
			secret, err := p.parseSecret()
			if err != nil {
				return nil, err
			}
			return secret, p.readRune()
		default:
			return character(p.current), nil
		}
	case token.Backslash:
		if token.IsToken(p.next) {
			r := character(p.next)
			return r, p.readRune()
		} else {
			return character(p.current), nil
		}
	default:
		return character(p.current), nil
	}
}
```

Reply:
Agree that a `return` can be more intuitive and robust than `continue`s in a for loop.

Improved a bit further, as the `return character(p.current), nil` can be re-used.

Reply:
5eb4c0

Reply:
See 5eb4c0b80cf9ace20e1032692fd5b19d9018135e